test/functional/p2p_outbound_eviction.py

```diff
@@ -0,0 +1,247 @@
+        node.setmocktime(cur_mock_time)
+        test_node.sync_with_ping()
+        # Wait for over 2 more min to trigger the disconnection
+        test_node.wait_for_getheaders()
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
+        node.setmocktime(cur_mock_time)
+        self.log.info("Test that the peer gets evicted")
+        test_node.wait_for_disconnect()
+
+        self.log.info("Create an outbound connection and send header but never catch up")
+        # Mimic a node that just falls behind for long enough
+        # This should also apply for a node doing IBD that does not catch up in time
+        # Get our tip header and its parent
+        tip_header = from_hex(CBlockHeader(), node.getblockheader(node.getbestblockhash(), False))
+        prev_header = from_hex(CBlockHeader(), node.getblockheader(f"{tip_header.hashPrevBlock:064x}", False))
+
+        # Connect a peer and make it send us headers ending in our tip's parent
+        test_node = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="outbound-full-relay")
+        test_node.send_and_ping(msg_headers([prev_header]))
+
+        # Trigger the timeouts
+        cur_mock_time += (CHAIN_SYNC_TIMEOUT + 1)
+        node.setmocktime(cur_mock_time)
+        test_node.sync_with_ping()
+        test_node.wait_for_getheaders()
+        cur_mock_time += (HEADERS_RESPONSE_TIME + 1)
+        node.setmocktime(cur_mock_time)
+        self.log.info("Test that the peer gets evicted")
+        test_node.wait_for_disconnect()
+
+        self.log.info("Create an outbound connection and keep lagging behind, but not too much")
+        # Test that if the peer never catches up with our current tip, but it does with the
+        # expected work that we set when setting the timer (that is, our tip at the time)
+        # we do not disconnect the peer
+        test_node = node.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, connection_type="outbound-full-relay")
+        self.log.info("Mine a block so our peer starts lagging")
+        best_block_hash = self.generateblock(node, output="raw(42)", transactions=[])["hash"]
+        tip_header = from_hex(CBlockHeader(), node.getblockheader(best_block_hash, False))
+        test_node.sync_with_ping()
+        self.log.info("Keep catching up with the old tip and check that we are not evicted")
+        for _ in range(10):
+            # Advance time but not enough to evict the peer
+            best_block_hash = self.generateblock(node, output="raw(42)", transactions=[])["hash"]
+            cur_mock_time += (CHAIN_SYNC_TIMEOUT + 1)
+            node.setmocktime(cur_mock_time)
+
+            # Send a header with the old tip
+            test_node.send_and_ping(msg_headers([tip_header]))
+            tip_header = from_hex(CBlockHeader(), node.getblockheader(best_block_hash, False))
+
+            # Check that we are not evicted
+            test_node.wait_for_getheaders()
```

Comment:
Ohh, TIL, I'm guessing this is because the collection is already populated with the `getheaders` from the interaction `i-1` and we find it all the time

Reply:
yeah! follow up idea, out of scope for this PR - it'd be awesome to modify `wait_for_getheaders()` to make sure that it's the same getheaders (which was sent from TestNode's `MaybeSendGetHeaders`) we're expecting that is received!

Reply:
oh see #29318.

Reply:
Fixed it by popping from `last_message`, may open a followup simplifying it after #https://github.com/bitcoin/bitcoin/issues/18614 gets fixed